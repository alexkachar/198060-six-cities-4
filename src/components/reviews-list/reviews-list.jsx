import React from 'react';
import PropTypes from 'prop-types';
import Review from '../review/review.jsx';
import ReviewTypes from '../../types/review.js';
import {connect} from 'react-redux';

import {AuthorizationStatus} from '../../store/reducers/user/user.js';
import {getAuthorizationStatus} from '../../store/reducers/user/selectors.js';
import ReviewOperation from '../../store/operations/review/review.js';
import ReviewForm from '../review-form/review-form.jsx';

import {store} from '../../index.js';
import {getReviews} from '../../store/reducers/review/selectors.js';

class ReviewsList extends React.PureComponent {

  componentDidMount() {
    const {offerId} = this.props;
    store.dispatch(ReviewOperation.loadReviews(offerId));
  }

  render() {
    const {offerId, reviews} = this.props;
    const reviewsCount = this.props.reviews.length;
    const isAuthorized = this.props.authorizationStatus === AuthorizationStatus.AUTH;

    return (
      <section className="property__reviews reviews">
        <h2 className="reviews__title">Reviews Â· <span className="reviews__amount">{reviewsCount}</span></h2>

        <ul className="reviews__list">
          {reviews.map((review) => <Review review={review} key={review.id} />)}
        </ul>

        {isAuthorized ? <ReviewForm offerId={offerId} /> : null}

      </section>
    );
  }

}

const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state),
  reviews: getReviews(state)
});

export {ReviewsList};

export default connect(mapStateToProps, null)(ReviewsList);

ReviewsList.propTypes = {
  reviews: PropTypes.arrayOf(ReviewTypes.isRequired).isRequired,
  authorizationStatus: PropTypes.string.isRequired,
  offerId: PropTypes.number.isRequired,
};
