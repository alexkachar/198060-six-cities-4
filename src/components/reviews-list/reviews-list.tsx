import * as React from 'react';


import ReviewCard from '../review/review-card';
import ReviewForm from '../review-form/review-form';
import Review from '../../interfaces/review';

interface Props {
  reviews: Review[];
  isAuthorized: boolean;
  offerId: number;
}

const ReviewsList = (props: Props) => {
  const {offerId, reviews, isAuthorized} = props;

  const reviewsCount = reviews.length;

  return (
    <section className="property__reviews reviews">
      <h2 className="reviews__title">Reviews Â· <span className="reviews__amount">{reviewsCount}</span></h2>

      <ul className="reviews__list">
        {reviews.map((review) => <ReviewCard review={review} key={review.id} />)}
      </ul>

      {isAuthorized ? <ReviewForm offerId={offerId} /> : null}

    </section>
  );
};

export default ReviewsList;
