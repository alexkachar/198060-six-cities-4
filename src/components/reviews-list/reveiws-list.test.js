import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import {ReviewsList} from './reviews-list.jsx';
import mockStore from '../../mocks/mock-store.js';

const mockReviews = [
  {
    id: 1,
    name: `Max`,
    isPro: false,
    userPic: `img/avatar-max.jpg`,
    rating: 4,
    text: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
    date: `2019-04-24`
  },
  {
    id: 2,
    name: `Angelina`,
    userPic: `img/avatar-angelina.jpg`,
    isPro: false,
    rating: 5,
    text: `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
    date: `2019-05-14`
  }
];

it(`ReviewsList renders correctly`, () => {
  const tree = renderer
  .create(
      <Provider store={mockStore}>
        <ReviewsList
          offerId={1}
          authorizationStatus={`AUTH`}
          reviews={mockReviews}
          onRequestReviews={() => mockReviews}
        />
      </Provider>, {
        createNodeMock: () => document.createElement(`div`)
      })
.toJSON();
  expect(tree).toMatchSnapshot();
});
