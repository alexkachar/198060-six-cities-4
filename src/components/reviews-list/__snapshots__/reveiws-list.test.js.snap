// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewsList renders correctly 1`] = `
<section
  className="property__reviews reviews"
>
  <h2
    className="reviews__title"
  >
    Reviews Â· 
    <span
      className="reviews__amount"
    >
      2
    </span>
  </h2>
  <ul
    className="reviews__list"
  >
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Max avatar"
            className="reviews__avatar user__avatar"
            height={54}
            src="img/avatar-max.jpg"
            width={54}
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Max
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "80%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              4
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.
        </p>
        <time
          className="reviews__time"
          dateTime="2019-04-24"
        >
          April 24, 2019
        </time>
      </div>
    </li>
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Angelina avatar"
            className="reviews__avatar user__avatar"
            height={54}
            src="img/avatar-angelina.jpg"
            width={54}
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Angelina
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "100%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              5
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.
        </p>
        <time
          className="reviews__time"
          dateTime="2019-05-14"
        >
          May 14, 2019
        </time>
      </div>
    </li>
  </ul>
  <form
    action="#"
    className="reviews__form form"
    method="post"
    onSubmit={[Function]}
  >
    <label
      className="reviews__label form__label"
      htmlFor="review"
    >
      Your review
    </label>
    <div
      className="reviews__rating-form form__rating"
    >
      <input
        className="form__rating-input visually-hidden"
        defaultValue={5}
        id="5-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="5-stars"
        title="perfect"
      >
        <svg
          className="form__star-image"
          height={33}
          width={37}
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        className="form__rating-input visually-hidden"
        defaultValue={4}
        id="4-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="4-stars"
        title="good"
      >
        <svg
          className="form__star-image"
          height={33}
          width={37}
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        className="form__rating-input visually-hidden"
        defaultValue={3}
        id="3-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="3-stars"
        title="not good"
      >
        <svg
          className="form__star-image"
          height={33}
          width={37}
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        className="form__rating-input visually-hidden"
        defaultValue={2}
        id="2-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="2-stars"
        title="badly"
      >
        <svg
          className="form__star-image"
          height={33}
          width={37}
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        className="form__rating-input visually-hidden"
        defaultValue={1}
        id="1-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="1-stars"
        title="terribly"
      >
        <svg
          className="form__star-image"
          height={33}
          width={37}
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
    </div>
    <textarea
      className="reviews__textarea form__textarea"
      defaultValue=""
      id="review"
      maxLength={300}
      name="review"
      onChange={[Function]}
      placeholder="Tell how was your stay, what you like and what can be improved"
    />
    <div
      className="reviews__button-wrapper"
    >
      <p
        className="reviews__help"
      >
        To submit review please make sure to set 
        <span
          className="reviews__star"
        >
          rating
        </span>
         and describe your stay with at least 
        <b
          className="reviews__text-amount"
        >
          50 characters
        </b>
        . Maximum review length is 
        <b
          className="reviews__text-amount"
        >
          300 characters
        </b>
        .
      </p>
      <button
        className="reviews__submit form__submit button"
        disabled={true}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</section>
`;
